File: Akshit_22b2531/week2/room_allocation.py
import heapq

def solve():
    n = int(input())
    customers = []
    
    for i in range(n):
        arrival, departure = map(int, input().split())
        # Store customer index for output order
        customers.append((arrival, departure, i))
    
    # Sort by arrival time
    customers.sort()
    
    rooms = []  # Min heap of (departure_time, room_number)
    room_assignments = [0] * n
    next_room = 1
    
    for arrival, departure, idx in customers:
        # Check if any room is free (has departure time <= current arrival)
        if rooms and rooms[0][0] <= arrival:
            # Reuse this room
            _, room_num = heapq.heappop(rooms)
            heapq.heappush(rooms, (departure, room_num))
            room_assignments[idx] = room_num
        else:
            # Need a new room
            heapq.heappush(rooms, (departure, next_room))
            room_assignments[idx] = next_room
            next_room += 1
    
    print(len(rooms))
    print(*room_assignments)

if __name__ == "__main__":
    solve() 

File: KRISHNA_24b2129/week2/Room Allocation.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

struct Customer {
    int arrival;
    int departure;
    int index;
};

bool compare(Customer a, Customer b) {
    return a.arrival < b.arrival;
}

int main() {
    int n;
    cin >> n;

    vector<Customer> customers(n);
    for (int i = 0; i < n; ++i) {
        cin >> customers[i].arrival >> customers[i].departure;
        customers[i].index = i;
    }

    sort(customers.begin(), customers.end(), compare);

    vector<int> room_endings; 
    vector<int> answer(n);

    for (int i = 0; i < n; ++i) {
        int assigned = -1;
        
        for (int j = 0; j < room_endings.size(); ++j) {
            if (room_endings[j] < customers[i].arrival) {
                assigned = j;
                break;
            }
        }

        if (assigned == -1) {
          
            room_endings.push_back(customers[i].departure);
            answer[customers[i].index] = room_endings.size(); 
        } else {
          
            room_endings[assigned] = customers[i].departure;
            answer[customers[i].index] = assigned + 1; 
        }
    }

    cout << room_endings.size() <<endl;
    for (int i = 0; i < n; ++i) {
        cout << answer[i] << " ";
    }
}